### PR TITLE
Fixing ordering of handlers in SD

### LINKED ISSUE #570: Sequence Diagram misplaces Handlers
## Symptoms

The Sequence Diagram misplaces the handler boxes in occasions and that results the message to appear out of order.

screen1
## Who's affected

This update is recommended for all users of ServiceInsight who rely on Sequence Diagram as it draws a more accurate sequence of messages in certain situations.

### PR DESCRIPTION
**Pull request by HEskandari**
~This would fix #570. This is a WIP.~

Fixes #570.

### CODE REVIEW

**Comment by HEskandari** [comment]
@Particular/serviceinsight-maintainers Please review.

**Comment by distantcam** [comment]
PR looks fine. I guess close the issue too?

File: src/ServiceInsight.Tests/ServiceInsight.Tests.csproj [hunk 1]
```diff
@@ -167,8 +169,5 @@
       <Name>ServiceInsight</Name>
     </ProjectReference>
   </ItemGroup>
-  <ItemGroup>
-    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
-  </ItemGroup>
```
Comment (adamralph) [inline src/ServiceInsight.Tests/ServiceInsight.Tests.csproj]: IIRC this speeds up Visual Studio's test discovery. Is there a reason it's been removed?

File: src/ServiceInsight/MessageViewers/XmlViewer/XmlMessageView.xaml.cs [hunk 2]
```diff
@@ -1,6 +1,5 @@
 ﻿namespace ServiceInsight.MessageViewers.XmlViewer
 {
-    using System;
```
Comment (adamralph) [inline src/ServiceInsight/MessageViewers/XmlViewer/XmlMessageView.xaml.cs]: I'm OK to merge as-is, but this really be in a separate commit.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs [hunk 3]
```diff
@@ -4,6 +4,7 @@ namespace ServiceInsight.SequenceDiagram
     using System.Collections.Generic;
     using System.Collections.ObjectModel;
     using System.Linq;
+    using DevExpress.Data.Filtering.Helpers;
```
Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: This has been added for no reason. It's unused.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs [hunk 4]
```diff
@@ -25,6 +26,7 @@ public ModelCreator(List<StoredMessage> messages, IMessageCommandContainer conta
             this.container = container;
 
             Initialize();
+            OrderHandlers();
```
Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: I'm not really keen on the introduction of a second method which reassigns the `handlers` field. Can we encapsulate the ordering inside `Initialize()` and avoid the reassignment?

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs [hunk 5]
```diff
@@ -100,6 +102,17 @@ void Initialize()
             }
         }
 
+        private void OrderHandlers()
+        {
+            var start = handlers.Where(h => h.ID == ConversationStartHandlerName);
```
Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: Why do we have to special case the conversation start? Is this because `HandledAt` is not granular enough and the conversation start can have the same value as the next handler?

File: src/ServiceInsight.Tests/ServiceInsight.Tests.csproj (hunk 1 above)
Comment (HEskandari) [inline src/ServiceInsight.Tests/ServiceInsight.Tests.csproj]: Not really. VS is just haphazardly removes or adds it. I can undo this change.

File: src/ServiceInsight/MessageViewers/XmlViewer/XmlMessageView.xaml.cs (hunk 2 above)
Comment (HEskandari) [inline src/ServiceInsight/MessageViewers/XmlViewer/XmlMessageView.xaml.cs]: I'll remove.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs (hunk 3 above)
Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: I'll remove.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs (hunk 5 above)
Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: this is about ordering to the best of our knowledge as the endpoints time could be out of sync. We know that logically the start of the conversation happens first, that's why it was special cased.

Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: OK, makes sense.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs (hunk 4 above)
Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: The main thing is that `Initialize` is doing a few things, including a first-step ordering. My thinking was that doing it as a part of Initialize (also a generalized name) would not making the second-step ordering obvious enough.

Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: What if `Initialize` created a list as a local variable and then assigned the `handlers` fields at the end, using the ordering? (The main code-smell is the reassignment of the `handlers` field.)

Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: okay, I understand now. Let me give it another go.

Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: @adamralph Please review again. Think I've addressed all the issues.

File: src/ServiceInsight/SequenceDiagram/ModelCreator.cs [hunk 6]
```diff
@@ -98,6 +99,19 @@ void Initialize()
 
                 sendingHandler.Out = sendingHandler.Out.Concat(new[] { arrow }).OrderBy(a => a).ToList();
             }
+
+            ReOrderHandlers(firstOrderHandlers);
+        }
+
+        private void ReOrderHandlers(IList<Handler> orderedHandlers)
```
Comment (adamralph) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: This method has mixed responsibilities:

- order the handlers
- assign the `handlers` field

Can we remove this method and inline the code in `Initialize`? It's only 4 statements, so I'm having a hard time justifying its existence.

Comment (distantcam) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: I disagree with this assessment.

The method is badly named, and could take an `IEnumerable` instead of an `IList`, which would make it more clear that it's assigning the handlers field with the values from the parameter. I don't think that's a mixed responsibility.

I'm also against inlining it. `Initialize` is long enough. It makes the code easier to understand, so there's the justification for it's existence.

Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: I was thinking the same way @distantcam. I think maybe `Initialize` is not a good name, and we should split that, not adding to the problem. It is already too large to comprehend.

Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: Let's discuss on our sync tomorrow.

Comment (HEskandari) [inline src/ServiceInsight/SequenceDiagram/ModelCreator.cs]: @adamralph @distantcam as discussed, for now I have merged the ordering method with initialize.

**Comment by distantcam** [comment]
@adamralph @HEskandari I thought the plan was to refactor the initialise method in another commit before this PR was ready. Is that not the case?

**Comment by HEskandari** [comment]
@distantcam My understanding was that we should do it separate as a Refactoring PR. I'll submit that separately.

**Comment by distantcam** [comment]
I'm -1 for a separate PR. I'm +1 for another commit to this PR.

**Comment by adamralph** [comment]
IIRC, when we discussed, this I did indeed suggest "separate refactoring commit" rather than "separate refactoring PR".

After reviewing the new diff, I can see that the this change is relatively trivial (a few extra lines in the method) so in this case I'd prefer that we merge the PR as-is (after a squash) so that the bug is fixed and we can release the patch. The refactoring is likely to have a significantly more complex diff, and I expect we'll have a few opinions on it, so I'd rather deal with that asynchronously.

**Comment by distantcam** [comment]
> The refactoring is likely to have a significantly more complex diff, and I expect we'll have a few opinions on it, so I'd rather deal with that asynchronously.

@adamralph Fair enough. Did you want to approve this PR then?

**Comment by HEskandari** [comment]
I'll squash the commits and update the PR.

**Comment by HEskandari** [comment]
@Particular/serviceinsight-maintainers The commits are merged. Please have one last review and merge if it's okay.
